bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+        // keep all ciphers
+        allAvailableStorageCiphers.add(storageCipher);
+    }
+
+    protected synchronized void unsetStorageCipher(StorageCipher storageCipher) {
+        allAvailableStorageCiphers.remove(storageCipher);
+        if (this.storageCipher.isPresent() && this.storageCipher.get() == storageCipher) {
+            this.storageCipher = Optional.empty();
+        }
+    }
+
+    private boolean isExpired(Date lastUsed) {
+        if (lastUsed == null) {
+            return false;
+        }
+        return System.currentTimeMillis() - lastUsed.getTime() > TimeUnit.MILLISECONDS.convert(EXPIRE_DAYS,
+                TimeUnit.DAYS);
+    }
+
+    private class StorageEntryRemovalListener implements RemovalListener<String, AccessTokenResponse> {
+        @Override
+        public void onRemoval(RemovalNotification<String, AccessTokenResponse> notification) {
+            String handle = notification.getKey();
+            // remove token, last use, and configuration too
+            storageFacade.removeByHandle(handle);
+        }
+    }
+
+    /**
+     * This is designed to simplify all the locking required for the store,
+     * which may get removed/ destroyed/ unavailable due to OSGI policy.
+     */
+    private class StorageFacade implements AutoCloseable {
+        private volatile Storage<String> _storage;
+        private final ReentrantLock storageLock = new ReentrantLock(); // for all operations on the storage
+        private final Gson gson;
+
+        public StorageFacade(@Nullable Storage<String> _storage) {
+            this._storage = _storage;
+
+            // Need to override the date serialization because the milli-seconds are truncated if not handled properly
+            gson = new GsonBuilder()
+                    .registerTypeAdapter(Date.class,
+                            (JsonDeserializer<Date>) (json, typeOfT,
+                                    context) -> new Date(json.getAsJsonPrimitive().getAsLong()))
+                    .registerTypeAdapter(Date.class, (JsonSerializer<Date>) (date, type,
+                            jsonSerializationContext) -> new JsonPrimitive(date.getTime()))
+                    .create();
+        }
+
+        @SuppressWarnings("unchecked")
+        public HashSet<String> getAllHandlesFromIndex() {
+            HashSet<String> handlesFromStoreageIndex = new HashSet<>();
+            try {
+                String allHandlesStr = get(STORE_KEY_INDEX_OF_HANDLES);
+                logger.debug("All available handles: {}", allHandlesStr);
+                if (allHandlesStr == null) {
+                    return handlesFromStoreageIndex;
+                }
+                return gson.fromJson(allHandlesStr, HashSet.class);
+            } catch (RuntimeException storeNotAvailable) {
+                return handlesFromStoreageIndex; // empty
+            }
+        }
+
+        public String get(@NonNull String key) {
+            try {
+                storageLock.lock();
```

Comment:
outside try block

Reply:
It is a recommended practice to lock the monitor outside try block. If you use it inside and if the lock fails, the finally will be invoked and it will again try to unlock the monitor which has never been locked.